### ISSUE TITLE
not authorized to perform: iam:ListMFADevices

### ISSUE THREAD

**Issue by justinffs** [lonelyplanet/aws-mfa#19]
A client error (AccessDenied) occurred when calling the ListMFADevices operation: User: arn:aws:iam::###########:user/XXXXXXXX is not authorized to perform: iam:ListMFADevices

**Comment by leejones** [comment]
@justinffs I think so. Do you have any MFA devices for your user?

It looks like the related logic is [here](https://github.com/lonelyplanet/aws-mfa/blob/master/lib/aws_mfa.rb#L51-L56). If the file `~/.aws/mfa_device` is present `aws-mfa` will use that, otherwise it calls to the AWS API (which is where you ran into permissions errors I think).

You will need the [ARN](http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) for the MFA device you want to use. The ARN will look something like `arn:aws:iam::account-id:mfa/virtual-device-name` as shown in the [ARN example docs](http://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arn-syntax-iam). Then you can add the ARN to a file at `~/.aws/mfa_device` and it should work.

**Comment by skemper** [comment]
Just hit this error with a user this morning.  Looks like it might be possible to add an optional `--user <username>` parameter to the call on [https://github.com/lonelyplanet/aws-mfa/blob/master/lib/aws_mfa.rb#L75](line 75).  At least for the AWS permissions we have set up, it would solve the problem.  Specifically, our current policy doesn't allow MFA users to list *all* devices, just their own.

@justinffs, if you run a `aws iam list-mfa-devices --user <your username>` using your regular API keys, do you get data back?  If so, this change might fix your troubles as well.

**Comment by justinffs** [comment]
I will check when I can and follow up with you.
I don't think I even have permission to see my own IAM information.

**Comment by sciurus** [comment]
@skemper that shouldn''t necessary, by default `list-mfa-devices` only list the devices that belong to the user calling it.

https://docs.aws.amazon.com/cli/latest/reference/iam/list-mfa-devices.html

**Comment by sciurus** [comment]
@justinffs if you can't see your own IAM information, then this isn't an issue with aws-mfa. Whoever administers your AWS account needs to give you permission to perform `list-mfa-devices` or give you the ARN of your MFA device.

**Comment by skemper** [comment]
@sciurus correct, but our AWS permissions don't allow an IAM account to call list-mfa-devices without specifying a particular device (their own).  We don't want IAM users to be able to discover any information about the MFA setup of other users.

**Comment by leejones** [comment]
One thing to note:

> If the request includes a IAM user name, then this action lists all the MFA devices associated with the specified user. If you do not specify a user name, IAM determines the user name implicitly based on the AWS access key ID signing the request for this API.

-- http://docs.aws.amazon.com/cli/latest/reference/iam/list-mfa-devices.html

It is possible to limit someone's permissions so they can only list their MFA devices. The policy might look something like:

```
...
   "Effect": "Allow",
      "Action": [
        "iam:ListMFADevices"
      ],
      "Resource": "arn:aws:iam::account-id-without-hyphens:mfa/${aws:username}"
...
```

You would want to research and adapt that to your context, but I mainly wanted to show that it's possible to have a variable for username in the policy.

**Comment by jstruzik** [comment]
I actually ran into this same problem recently even with the lenient IAM permissions applied:

```json
...
   "Effect": "Allow",
      "Action": [
        "iam:ListMFADevices"
      ],
      "Resource": "arn:aws:iam::accountid:mfa/*
...
```

The strange part was running [`list-mfa-devices`](https://docs.aws.amazon.com/cli/latest/reference/iam/list-mfa-devices.html) should pick up my user implicitly but it seemed to give me a permission error. However, adding the `--user-name` option for my own user seemed to have worked.

I'm not sure if this is a bug on AWS's CLI/API side or not but for now I circumvented the issue by following the steps in https://github.com/lonelyplanet/aws-mfa/issues/19#issuecomment-215435092.

https://github.com/lonelyplanet/aws-mfa/pull/28 might be a good solution for this issue but that also adds an IAM dependency of `get-user` for the authenticated entity.

**Comment by ksperling** [comment]
Instead of using `get-user` the user id can be derived by calling `aws sts get-caller-identity`, which should be much less of a security issues because it is limited to the caller by design.

The policy for ListMfaDevices that I'm using is

        {
            "Action": "iam:ListMfaDevices",
            "Effect": "Allow",
            "Resource": "arn:aws:iam::ACCOUNT:user/${aws:username}"
        }

I.e. the resource is the user rather than the MFA device itself

**Comment by alsmola** [comment]
I was having this same issue, and I was able to get it working by adding the following policy for the user attempting to run the command:

        {
            "Effect": "Allow",
            "Action": "iam:ListMFADevices",
            "Resource": "*"
        }

I started with a `Resource` attribute of `arn:aws:iam::ACCOUNT:user/${aws:username}`, moved to `arn:aws:iam::ACCOUNT:user/*`, but neither worked. I tried specifying the resource attribute more and more permissively, like `arn:aws:iam::*:user/*`, tried adding `arn:aws:iam::*:mfa/*`, etc. but the full wildcard was the only thing that worked.

Note that I was able to verify the same functionality when using the aws cli.
`aws iam list-mfa-devices --user-name MYUSERNAME` worked for the more restrictive policies, whereas `aws iam list-mfa-devices` did not until I specified the full wildcard policy.

**Comment by ksperling** [comment]
I think the wildcard policy is not great from a security point of view, as it allows an attacker to pin-point users that don't have MFA enabled. With the PRs that are open, the restrictive `${aws:username}` version of the policy works with `aws-iam`.

I'm currently using a slight variation that seems to work for allowing the case where no user is specified to work as well:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "iam:ListMfaDevices",
            "Resource": "arn:aws:iam::*:user/${aws:username}"
        },
        {
            "Sid": "AllowListMfaDevicesWithoutExplicitResource",
            "Effect": "Allow",
            "Action": "iam:ListMfaDevices",
            "NotResource": "arn:aws:iam::*"
        }
    ]
}
```

**Comment by phumpal** [comment]
>             "NotResource": "arn:aws:iam::*"

Is this valid without an arn?
